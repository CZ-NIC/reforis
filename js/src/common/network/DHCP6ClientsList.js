/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import moment from "moment";
import { formFieldsSize } from "foris";

DHCP6ClientsList.propTypes = {
    ipv6clients: PropTypes.arrayOf(PropTypes.object),
};

export default function DHCP6ClientsList({ ipv6clients }) {
    return (
        <div className={formFieldsSize}>
            <h2>{_("IPv6 DHCP Client List")}</h2>
            <p>
                {_(
                    "This list contains all devices that are connected to the network through wired or wireless connections using IPv6."
                )}
            </p>
            {ipv6clients.length === 0 ? (
                <p className="text-muted text-center">
                    {_("No clients found.")}
                </p>
            ) : (
                <div className="table-responsive">
                    <table className="table table-hover">
                        <thead className="thead-light">
                            <tr className="text-left">
                                <th>{_("Hostname")}</th>
                                <th>{_("IPv6 Address")}</th>
                                <th>{_("DUID")}</th>
                                <th className="text-center">{_("Expires")}</th>
                            </tr>
                        </thead>
                        <tbody>
                            {ipv6clients.map((client) => (
                                <DHCP6ClientsListItem
                                    key={client.ipv6}
                                    {...client}
                                />
                            ))}
                        </tbody>
                    </table>
                </div>
            )}
        </div>
    );
}

DHCP6ClientsListItem.propTypes = {
    ipv6: PropTypes.string.isRequired,
    expires: PropTypes.number.isRequired,
    duid: PropTypes.string.isRequired,
    hostname: PropTypes.string.isRequired,
};

function DHCP6ClientsListItem({ ipv6, expires, duid, hostname }) {
    return (
        <tr className="text-left">
            <td>{hostname}</td>
            <td>{ipv6}</td>
            <td>{duid}</td>
            <td className="text-center">
                {moment.unix(expires).format("YYYY-MM-DD HH:mm")}
            </td>
        </tr>
    );
}
