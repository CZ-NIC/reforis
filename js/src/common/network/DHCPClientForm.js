/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

import TextInput from '../bootstrap/TextInput';
import {validateDomain} from '../validations';

const HELP_TEXTS = {
    hostname: _('Hostname which will be provided to DHCP server.'),
};

DHCPClientForm.propTypes = {
    formData: propTypes.shape({
        hostname: propTypes.string,
    }).isRequired,
    formErrors: propTypes.shape({
        hostname: propTypes.string
    }),
    setFormValue: propTypes.func.isRequired,
    updateRule: propTypes.func.isRequired,
};

export default function DHCPClientForm({formData, formErrors, setFormValue, updateRule, ...props}) {
    return <TextInput
        label={_('DHCP hostname')}
        value={formData.hostname || ''}
        error={(formErrors || {}).hostname || null}
        helpText={HELP_TEXTS.hostname}

        onChange={setFormValue(
            value => updateRule({hostname: {$set: value}})
        )}

        {...props}
    />
}

export function validateDHCPForm(formData) {
    const error = {hostname: validateDomain(formData.hostname)};
    return error.hostname ? error : null;
}
