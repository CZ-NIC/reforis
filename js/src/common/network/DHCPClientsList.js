/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import moment from "moment";

DHCPClientsList.propTypes = {
    clients: PropTypes.arrayOf(PropTypes.object),
};

export default function DHCPClientsList({ clients }) {
    return (
            <h2>{_("DHCP Client List")}</h2>
            <p>
                {_(
                    "This list contains all devices that are connected to the network through wired or wireless connections."
                )}
            </p>
            {clients.length === 0 ? (
                <p className="text-muted text-center">
                    {_("No DHCP clients found.")}
                </p>
            ) : (
                <div className="table-responsive">
                    <table className="table table-hover">
                        <thead>
                            <tr className="text-center">
                                <th>{_("Expires")}</th>
                                <th>{_("IP Address")}</th>
                                <th>{_("MAC Address")}</th>
                                <th>{_("Hostname")}</th>
                                <th>{_("Active")}</th>
                            </tr>
                        </thead>
                        <tbody>
                            {clients.map((client) => (
                                <DHCPClientsListItem
                                    key={client.ip}
                                    {...client}
                                />
                            ))}
                        </tbody>
                    </table>
                </div>
            )}
        </>
    );
}

DHCPClientsListItem.propTypes = {
    ip: PropTypes.string.isRequired,
    expires: PropTypes.number.isRequired,
    mac: PropTypes.string.isRequired,
    hostname: PropTypes.string.isRequired,
    active: PropTypes.bool.isRequired,
};

function DHCPClientsListItem({ ip, expires, mac, hostname, active }) {
    return (
        <tr className="text-center">
            <td>{moment.unix(expires).format("YYYY-MM-DD HH:mm")}</td>
            <td>{ip}</td>
            <td>{mac}</td>
            <td>{hostname}</td>
            <td>
                <i className={`fas ${active ? "fa-check" : "fa-times"}`} />
            </td>
        </tr>
    );
}
