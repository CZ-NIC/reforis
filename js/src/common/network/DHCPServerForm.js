/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

import NumberInput from 'common/bootstrap/NumberInput';

export const HELP_TEXT = _('Enable this option to automatically assign IP addresses to the devices connected to the router.');

DHCPServerForm.propTypes = {
    formData: propTypes.shape({
        start: propTypes.number,
        limit: propTypes.number,
        lease_time: propTypes.number,
    }).isRequired,
    setFormValue: propTypes.func.isRequired,
    updateRule: propTypes.func.isRequired,
};

export default function DHCPServerForm({formData, updateRule, setFormValue, ...props}) {
    return <>
        <NumberInput
            label={_('DHCP start')}
            value={formData.start}
            min='1'
            required

            onChange={setFormValue(
                value => updateRule({start: {$set: value}})
            )}

            {...props}
        />
        <NumberInput
            label={_('DHCP max leases')}
            value={formData.limit}
            min='1'
            required

            onChange={setFormValue(
                value => updateRule({limit: {$set: value}})
            )}

            {...props}
        />
        <NumberInput
            label={_('Lease time (hours)')}
            value={formData.lease_time}
            min='120'
            required

            onChange={setFormValue(
                value => updateRule({lease_time: {$set: value}})
            )}

            {...props}
        />
    </>
}
