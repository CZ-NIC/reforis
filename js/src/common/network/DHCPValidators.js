/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import ipaddr from "ipaddr";

import { undefinedIfEmpty, withoutUndefinedKeys, validateIPv4Address } from "foris";

import { validateRequiredField } from "./validators";
import { addToAddress } from "./utils";

/*
 * Check if DHCP start address is within network.
 * - networkAddress - address object (parsed by ipaddr)
 * - prefixLength - network prefix related to mask (e.g. 24 for 255.255.255.0)
 * - DHCPStartAddress - address object (parsed by ipaddr)
 */
function isDHCPStartWithinNetwork(networkAddress, prefixLength, DHCPStartAddress) {
    // Check if DHCP start address is within network
    return DHCPStartAddress.match(networkAddress, prefixLength);
}

/*
 * Check if DHCP start address clashes with a reserved address.
 * - networkAddress - address object (parsed by ipaddr)
 * - broadcastAddress - address object (parsed by ipaddr)
 * - DHCPStartAddress - address object (parsed by ipaddr)
 */
function isDHCPStartReserved(networkAddress, broadcastAddress, DHCPStartAddress) {
    // Check if DHCP start address is not at the borders of address range
    if ([broadcastAddress.toString(), networkAddress.toString()]
        .includes(DHCPStartAddress.toString())
    ) {
        return true;
    }
    return false;
}

/*
 * Check if number of leases can be realized in this network.
 * - networkAddress - address object (parsed by ipaddr)
 * - prefixLength - address object (parsed by ipaddr)
 * - DHCPStart - address object (parsed by ipaddr)
 * - maxLeases - maximum number of addresses used by DHCP
 */
function isMaxLeasesValid(networkAddress, prefixLength, DHCPStart, maxLeases) {
    const maxAddress = addToAddress(DHCPStart, maxLeases - 1);
    return maxAddress.match(networkAddress, prefixLength);
}

function validatePositiveNumber(number) {
    // eslint-disable-next-line no-restricted-globals
    if (number === "" || isNaN(number)) {
        return _("Value must be a number.");
    } if (number < 1) {
        return _("Value must be positive.");
    }
    return undefined;
}

export default function validateDHCP(
    dhcpSettings, routerIPString, networkMask, isNetworkConfigBroken,
) {
    if (isNetworkConfigBroken) {
        return {
            start: _("Invalid network settings. DHCP settings can't be validated."),
        };
    }

    const errors = {
        start: (
            validateRequiredField(dhcpSettings.start)
            || validateIPv4Address(dhcpSettings.start)
        ),
        limit: validatePositiveNumber(dhcpSettings.limit),
        lease_time: dhcpSettings.lease_time < 1 ? _("Minimum lease time is 1 hour.") : undefined,
    };

    if (!errors.start) {
        const prefixLength = ipaddr.IPv4.parse(networkMask).prefixLengthFromSubnetMask();
        const CIDR = `${routerIPString}/${prefixLength}`;
        const networkAddress = ipaddr.IPv4.networkAddressFromCIDR(CIDR);
        const broadcastAddress = ipaddr.IPv4.broadcastAddressFromCIDR(CIDR);
        const DHCPStart = ipaddr.IPv4.parse(dhcpSettings.start);

        if (!isDHCPStartWithinNetwork(networkAddress, prefixLength, DHCPStart)) {
            errors.start = _("Address is outside current network.");
        } else if (isDHCPStartReserved(networkAddress, broadcastAddress, DHCPStart)) {
            errors.start = _("Address is already reserved for other purposes.");
        } else if (!errors.limit) { // There are no more important errors so far.
            if (!isMaxLeasesValid(networkAddress, prefixLength, DHCPStart, dhcpSettings.limit)) {
                errors.limit = _("Too many addresses requested. Set a lower number or change DHCP start.");
            } else if (DHCPStart.toString() === routerIPString && dhcpSettings.limit === 1) {
                errors.limit = _("The only DHCP lease is the same as router's address. Increase limit or change start address.");
            }
        }
    }

    return undefinedIfEmpty(withoutUndefinedKeys(errors));
}
