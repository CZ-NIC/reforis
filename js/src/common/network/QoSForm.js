/*
 * Copyright (C) 2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import { NumberInput, Switch, undefinedIfEmpty } from "foris";

QoSForm.propTypes = {
    formData: PropTypes.shape({
        enabled: PropTypes.bool,
        download: PropTypes.string,
        upload: PropTypes.string,
    }).isRequired,
    setFormValue: PropTypes.func.isRequired,
    formErrors: PropTypes.shape({
        download: PropTypes.string,
        upload: PropTypes.string,
    }),
    helpTexts: PropTypes.shape({
        download: PropTypes.string,
        upload: PropTypes.string,
        enabled: PropTypes.string,
    }),
    disabled: PropTypes.bool,
};

QoSForm.defaultProps = {
    formErrors: {},
};

export default function QoSForm({
    formData,
    formErrors,
    setFormValue,
    disabled,
    helpTexts,
}) {
    return (
        <>
            <Switch
                label={_("Enable QoS")}
                checked={formData.enabled}
                helpText={helpTexts.enabled}
                onChange={setFormValue((value) => ({
                    qos: { enabled: { $set: value } },
                }))}
                disabled={disabled}
            />
            {formData.enabled && (
                <>
                    <NumberInput
                        label={_("Download")}
                        value={formData.download}
                        error={formErrors.download}
                        helpText={helpTexts.download}
                        inlineText="kb/s"
                        min="1"
                        required
                        onChange={setFormValue((value) => ({
                            qos: { download: { $set: value } },
                        }))}
                        disabled={disabled}
                    />
                    <NumberInput
                        label={_("Upload")}
                        value={formData.upload}
                        error={formErrors.upload}
                        helpText={helpTexts.upload}
                        inlineText="kb/s"
                        min="1"
                        required
                        onChange={setFormValue((value) => ({
                            qos: { upload: { $set: value } },
                        }))}
                        disabled={disabled}
                    />
                </>
            )}
        </>
    );
}

export function validateQoS(formData) {
    const errors = {};
    if (formData.download < 1) {
        errors.download = _("Download speed must be positive.");
    }
    if (formData.upload < 1) {
        errors.upload = _("Upload speed must be positive.");
    }
    return undefinedIfEmpty(errors);
}
