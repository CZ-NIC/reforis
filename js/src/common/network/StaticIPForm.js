/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import { TextInput, validateIPv4Address } from "foris";

const HELP_TEXTS = {
    dns: _(
        "DNS server address is not required as the built-in DNS resolver is capable of working without it."
    ),
};

const FIELDS_PROP_TYPES = {
    ip: PropTypes.string,
    netmask: PropTypes.string,
    gateway: PropTypes.string,
    dns1: PropTypes.string,
    dns2: PropTypes.string,
};

StaticIPForm.propTypes = {
    formData: PropTypes.shape(FIELDS_PROP_TYPES).isRequired,
    formErrors: PropTypes.shape(FIELDS_PROP_TYPES),
    setFormValue: PropTypes.func.isRequired,
    updateRule: PropTypes.func.isRequired,
    disabled: PropTypes.bool,
};

StaticIPForm.defaultProps = {
    formData: {},
    formErrors: {},
};

export default function StaticIPForm({
    formData,
    formErrors,
    updateRule,
    setFormValue,
    disabled,
}) {
    return (
        <>
            <TextInput
                label={_("IP address")}
                value={formData.ip || ""}
                error={formErrors.ip || null}
                required
                onChange={setFormValue((value) =>
                    updateRule({ ip: { $set: value } })
                )}
                disabled={disabled}
            />
            <TextInput
                label={_("Network mask")}
                value={formData.netmask || ""}
                error={formErrors.netmask || null}
                required
                onChange={setFormValue((value) =>
                    updateRule({ netmask: { $set: value } })
                )}
                disabled={disabled}
            />
            <TextInput
                label={_("Gateway")}
                value={formData.gateway || ""}
                error={formErrors.gateway || null}
                required
                onChange={setFormValue((value) =>
                    updateRule({ gateway: { $set: value } })
                )}
                disabled={disabled}
            />
            <TextInput
                label={_("DNS server 1")}
                value={formData.dns1 || ""}
                error={formErrors.dns1 || null}
                helpText={HELP_TEXTS.dns}
                onChange={setFormValue((value) =>
                    updateRule({ dns1: { $set: value } })
                )}
                disabled={disabled}
            />
            <TextInput
                label={_("DNS server 2")}
                value={formData.dns2 || ""}
                error={formErrors.dns2 || null}
                helpText={HELP_TEXTS.dns}
                onChange={setFormValue((value) =>
                    updateRule({ dns2: { $set: value } })
                )}
                disabled={disabled}
            />
        </>
    );
}

export function validateStaticForm(formData) {
    const errors = {};
    ["ip", "netmask", "gateway", "dns1", "dns2"].forEach((field) => {
        const error = validateIPv4Address(formData[field]);
        if (error) errors[field] = error;
    });
    ["ip", "netmask", "gateway"].forEach((field) => {
        if (!formData[field] || formData[field] === "")
            errors[field] = _("This field is required.");
    });

    return JSON.stringify(errors) !== "{}" ? errors : null;
}
