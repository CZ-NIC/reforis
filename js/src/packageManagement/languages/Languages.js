/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { ForisForm } from "foris";

import API_URLs from "common/API";
import DisableIfUpdaterIsDisabled from "../utils/DisableIfUpdaterIsDisabled";
import LanguagesForm from "./LanguagesForm";

export default function Languages() {
    return (
        <>
            <h1>{_("Languages")}</h1>
            <p>
                {_(
                    "If you want to use another language than English you can select it from the following list:"
                )}
            </p>
            <ForisForm
                forisConfig={{ endpoint: API_URLs.languagePackages }}
                prepDataToSubmit={prepDataToSubmit}
                validator={validator}
            >
                <DisableIfUpdaterIsDisabled>
                    <LanguagesForm />
                </DisableIfUpdaterIsDisabled>
            </ForisForm>
        </>
    );
}

function prepDataToSubmit(formData) {
    const languages = formData.languages
        .filter((language) => language.enabled)
        .map((language) => language.code);

    return { languages };
}

// Hack to disable submit button
function validator(formData) {
    if (!formData.enabled) return { enabled: true };
    return null;
}
