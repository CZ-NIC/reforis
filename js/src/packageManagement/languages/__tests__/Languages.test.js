/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import diffSnapshot from "snapshot-diff";

import {
    cleanup,
    render,
    wait,
    fireEvent,
} from "foris/testUtils/customTestRender";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from "jest-mock-axios";

import languagesFixture from "./__fixtures__/languages";
import Languages from "../Languages";

describe("<Packages/>", () => {
    let firstRender;
    let queryByText;
    let getByText;
    let getByLabelText;

    beforeEach(async () => {
        let asFragment;
        ({ queryByText, getByLabelText, getByText, asFragment } = render(
            <Languages />
        ));
        await wait(() =>
            expect(mockAxios.get).toBeCalledWith(
                "/reforis/api/language-packages",
                expect.anything()
            )
        );
        mockAxios.mockResponse({ data: languagesFixture(true) });
        await wait(() => getByText("CS"));
        firstRender = asFragment();
    });

    it("Should handle error.", async () => {
        const { getByText } = render(<Languages />);
        mockJSONError();
        await wait(() => {
            expect(
                getByText("An error occurred while fetching data.")
            ).toBeTruthy();
        });
    });

    it("Should render: updates enabled.", () => {
        expect(firstRender).toMatchSnapshot();
    });

    it("Should render: updates disabled.", async () => {
        cleanup();
        const { getByText, asFragment } = render(<Languages />);
        await wait(() =>
            expect(mockAxios.get).toBeCalledWith(
                "/reforis/api/language-packages",
                expect.anything()
            )
        );
        mockAxios.mockResponse({ data: languagesFixture(false) });

        await wait(() => getByText("CS"));

        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it("Send languages.", () => {
        fireEvent.click(getByText("Save"));
        expect(mockAxios.post).toHaveBeenCalledWith(
            "/reforis/api/language-packages",
            {
                languages: ["cs"],
            },
            expect.anything()
        );
    });
});
