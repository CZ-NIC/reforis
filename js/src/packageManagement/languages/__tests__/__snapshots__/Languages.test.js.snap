// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Packages/> Should render: updates disabled. 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -7,10 +7,26 @@
    </p>
    <div
      class=\\"card p-4 col-sm-12 col-lg-12 p-0 mb-4\\"
    >
      <form>
+       <div
+         class=\\"alert  alert-warning\\"
+       >
+         <span>
+           Please enable 
+           <a
+             href=\\"/reforis/package-management/update-settings\\"
+           >
+             automatic updates
+           </a>
+            to manage packages and languages.
+         </span>
+       </div>
+       <div
+         class=\\"text-muted\\"
+       >
          <h2>
            Languages List
          </h2>
          <div
            class=\\"container-fluid mb-3\\"
@@ -28,10 +44,11 @@
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      checked=\\"\\"
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"1\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -51,10 +68,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"2\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -74,10 +92,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"3\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -97,10 +116,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"4\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -120,10 +140,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"5\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -143,10 +164,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"6\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -166,10 +188,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"7\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -189,10 +212,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"8\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -212,10 +236,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"9\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -235,10 +260,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"10\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -258,10 +284,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"11\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -272,15 +299,17 @@
                  </div>
                </div>
              </div>
            </div>
          </div>
+       </div>
        <div
          class=\\"text-right\\"
        >
          <button
            class=\\"btn btn-primary  col-sm-12 col-md-3 col-lg-2\\"
+           disabled=\\"\\"
            type=\\"submit\\"
          >
            Save
          </button>
        </div>"
`;

exports[`<Packages/> Should render: updates enabled. 1`] = `
<DocumentFragment>
  <h1>
    Languages
  </h1>
  <p>
    If you want to use another language other than English you can select it from the following list:
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <form>
      <h2>
        Languages List
      </h2>
      <div
        class="container-fluid mb-3"
      >
        <div
          class="row justify-content-start"
        >
          <div
            class="col-2 col-lg-1 mr-4"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  checked=""
                  class="custom-control-input"
                  id="1"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="1"
                >
                  CS
                </label>
              </div>
            </div>
          </div>
          <div
            class="col-2 col-lg-1 mr-4"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="2"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="2"
                >
                  DA
                </label>
              </div>
            </div>
          </div>
          <div
            class="col-2 col-lg-1 mr-4"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="3"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="3"
                >
                  DE
                </label>
              </div>
            </div>
          </div>
          <div
            class="col-2 col-lg-1 mr-4"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="4"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="4"
                >
                  FR
                </label>
              </div>
            </div>
          </div>
          <div
            class="col-2 col-lg-1 mr-4"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="5"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="5"
                >
                  LT
                </label>
              </div>
            </div>
          </div>
          <div
            class="col-2 col-lg-1 mr-4"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="6"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="6"
                >
                  PL
                </label>
              </div>
            </div>
          </div>
          <div
            class="col-2 col-lg-1 mr-4"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="7"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="7"
                >
                  RU
                </label>
              </div>
            </div>
          </div>
          <div
            class="col-2 col-lg-1 mr-4"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="8"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="8"
                >
                  SK
                </label>
              </div>
            </div>
          </div>
          <div
            class="col-2 col-lg-1 mr-4"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="9"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="9"
                >
                  HU
                </label>
              </div>
            </div>
          </div>
          <div
            class="col-2 col-lg-1 mr-4"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="10"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="10"
                >
                  IT
                </label>
              </div>
            </div>
          </div>
          <div
            class="col-2 col-lg-1 mr-4"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="11"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="11"
                >
                  NB
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-md-3 col-lg-2"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
